codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,756 @@
+
+        checkNotNull(dst, "dst").setBytes(dstIdx, value, srcIdx, length);
+    }
+
+    /**
+     * Copies the content of this string to a {@link ByteBuf} using {@link ByteBuf#writeBytes(byte[], int, int)}.
+     *
+     * @param srcIdx the starting offset of characters to copy.
+     * @param dst the destination byte array.
+     * @param length the number of characters to copy.
+     */
+    public final void copy(int srcIdx, ByteBuf dst, int length) {
+        final byte[] value = this.value;
+        final int thisLen = value.length;
+
+        if (srcIdx < 0 || length > thisLen - srcIdx) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= srcIdx(" + srcIdx + ") <= srcIdx + length("
+                            + length + ") <= srcLen(" + thisLen + ')');
+        }
+
+        checkNotNull(dst, "dst").writeBytes(value, srcIdx, length);
+    }
+
+    /**
+     * Copies the content of this string to a byte array.
+     *
+     * @param srcIdx the starting offset of characters to copy.
+     * @param dst the destination byte array.
+     * @param dstIdx the starting offset in the destination byte array.
+     * @param length the number of characters to copy.
+     */
+    public final void copy(int srcIdx, byte[] dst, int dstIdx, int length) {
+        final byte[] value = this.value;
+        final int thisLen = value.length;
+
+        if (srcIdx < 0 || length > thisLen - srcIdx) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= srcIdx(" + srcIdx + ") <= srcIdx + length("
+                            + length + ") <= srcLen(" + thisLen + ')');
+        }
+
+        System.arraycopy(value, srcIdx, checkNotNull(dst, "dst"), dstIdx, length);
+    }
+
+    @Override
+    public int hashCode() {
+        if (hash != 0) {
+            return hash;
+        }
+
+        for (int i = 0; i < value.length; ++i) {
+            hash = hash * HASH_CODE_PRIME ^ value[i] & HASH_CODE_PRIME;
+        }
+
+        return hash;
+    }
+
+    /**
+     * Copies a range of characters into a new string.
+     *
+     * @param start the offset of the first character.
+     * @return a new string containing the characters from start to the end of the string.
+     * @throws IndexOutOfBoundsException if {@code start < 0} or {@code start > length()}.
+     */
+    public final ByteString subSequence(int start) {
+        return subSequence(start, length());
+    }
+
+    public ByteString subSequence(int start, int end) {
+        if (start < 0 || start > end || end > length()) {
+            throw new IndexOutOfBoundsException("expected: 0 <= start(" + start + ") <= end (" + end + ") <= length("
+                            + length() + ')');
+        }
+
+        if (start == 0 && end == value.length) {
+            return this;
+        }
+
+        if (end == start) {
+            return EMPTY_STRING;
+        }
+
+        return new ByteString(value, start, end - start, false);
+    }
+
+    public final int parseInt() {
```

Comment:
Yes the `parse*()` methods are essentially expecting ascii for the range being parsing.  While it is true that `ByteString` may be able to represent an `AsciiString` but the `AsciiString` guarantees a certain character encoding is enforced.  If we were to remove the `AsciiString` class we would no longer have an immutable `ByteString` that `implements CharSequence` (`ByteString` could implement something like `ByteSequence` but I don't think this exists).  Not implementing `CharSequence` makes it more difficult to use with other `CharSequence` objects (i.e. `String`).  Losing this interface prevents us from having a type of `ByteString` in algorithms which expect a `CharSequence` type. We may also miss out on a bunch of opportunities to use the `forEach(ByteProcessor ...)` interface which has been shown to be an improvement.

So in summary it can be done...but there are implications that we should consider before going this route.